source/core_mqtt_agent.c

```diff
@@ -621,9 +648,9 @@ static void handleAcks( const MQTTAgentContext_t * pAgentContext,
 
 static MQTTAgentContext_t * getAgentFromMQTTContext( MQTTContext_t * pMQTTContext )
 {
-    void * ret = pMQTTContext;
+    size_t offset = ( ( size_t ) &( ( ( MQTTAgentContext_t * ) 0 )->mqttContext ) );
 
-    return ( MQTTAgentContext_t * ) ret;
+    return ( MQTTAgentContext_t * ) &pMQTTContext[ 0U - offset ];
```

Comment:
Does this not have to be cast to uint8_t? Otherwise skipping offset*sizeof( MQTTContext_t * ) backwards. Also probably should not be dereferenced, as is relative to pointed to data rather than pointer itself?